File: Week 6/id_445/212.单词搜索-ii.cpp
/*
 * @lc app=leetcode.cn id=212 lang=cpp
 *
 * [212] 单词搜索 II
 */
#include <vector>
#include <string>
#include <map>
using namespace std;

// @lc code=start

struct Node
{
    bool word;
    string str;
    map<char, Node *> words;
};

class Trie
{
    Node *root;

public:
    Trie()
    {
        root = new Node();
    }
    void insert(string word)
    {
        Node *p = root;
        for (char w : word)
        {
            if (p->words.find(w) == p->words.end())
            {
                Node *n = new Node();
                p->words[w] = n;
            }
            p = p->words[w];
        }
        p->str = word;
        p->word = true;
    }
    void search(vector<string> & res,vector<vector<char>> & board){
        for (int i = 0; i < board.size(); i++)
        {
            for(int j=0;j<board[i].size();j++){
                help(res,board,root,i,j);
            }
        }
        
    }
    void help(vector<string> &res,vector<vector<char>> & board,Node *p,int x,int y){
        if(p->word){
            p->word=false;
            res.push_back(p->str);
            return;
        }
        if(x<0 || x==board.size() || y<0 ||y==board[x].size()) return;
        if(p->words.find(board[x][y])==p->words.end()) return;
        p=p->words[board[x][y]];
        char cur=board[x][y];
        board[x][y]='#';
        help(res,board,p,x+1,y);
        help(res,board,p,x-1,y);
        help(res,board,p,x,y+1);
        help(res,board,p,x,y-1);
        board[x][y]=cur;
    }
};
class Solution
{
public:
    vector<string> findWords(vector<vector<char>> &board, vector<string> &words)
    {
        Trie *t=new Trie();
        vector<string> res;
        for(string & s:words){
            t->insert(s);
        }
        t->search(res,board);
        return res;
    }
};
// @lc code=end

File: Week 6/id_520/212.单词搜索-ii.java
import java.util.LinkedList;
import java.util.List;

/*
 * @lc app=leetcode.cn id=212 lang=java
 *
 * [212] 单词搜索 II
 */

// @lc code=start
class Solution {

    class TrieNode {
        TrieNode[] children = new TrieNode[26];
        String word;
    }

    public List<String> findWords(char[][] board, String[] words) {
        List<String> result = new LinkedList<>();
        TrieNode root1 = buildTrie(words);
        for (int i = 0; i < board.length; i++) {
            for (int j = 0 ; j < board[0].length; j++) {
                dfs(board,i,j,root1,result);
            }
        }
        return result;
    }

    //相当于208题中的 insert
    public TrieNode buildTrie(String[] words) {
        //和208不同，这里每一个测试用例都需要单独建立一个Trie，208建立起来一个
        //然后被别的程序调用
        TrieNode root = new TrieNode();
        for (String s1 : words) {
            // root不能动，创建Trie的过程每次用node表示当前节点
            TrieNode node = root;
            for (char c : s1.toCharArray()) {
                if (node.children[c - 'a'] == null) node.children[c - 'a'] = new TrieNode();
                node = node.children[c - 'a'];
            }
            node.word = s1;// Store a word at the leaf node
        }
        return root;
    }

    //DFS
    public void dfs(char[][] board, int i, int j, TrieNode node, 
                    List<String> result) {
        char c = board[i][j];
        // terminator
        if (c == '#' || node.children[c - 'a'] == null) return;

        // process current logic
        node = node.children[c - 'a'];
        if (node.word != null) {  // found one
            result.add(node.word);
            node.word = null;   //de-duplicate
        }
        // '#' marked cell has been visited
        board[i][j] = '#';
        //left
        if (i > 0) dfs(board, i-1, j, node, result);
        //up
        if (j > 0) dfs(board, i, j-1, node, result);
        //right
        if (i < board.length - 1) dfs(board,i+1, j, node, result);
        //down
        if (j < board[0].length - 1) dfs(board,i, j+1, node, result);
        board[i][j] = c;
    }
}
// @lc code=end

